### PR TITLE
Improve extensions error handling

### LINKED ISSUE #3204: StrawberryGraphQLError in Custom Extension causes 500 error response instead of formatted error
## Description
I am attempting to add a Custom Extension to catch queries that are too large. I then want the extension to raise StrawberryGraphQLError and return a properly formatted error message. 

The issue is that when I raise StrawberryGraphQLError an internal error response is sent to the requester.

```
MAX_QUERY_LENGTH = 8192 # 8KB

class MaxQueryLengthExtension(SchemaExtension):
    async def on_operation(self):
        if len(self.execution_context.query) > MAX_QUERY_LENGTH:
            raise StrawberryGraphQLError(message="Query too large")
        yield
```

Stack Trace

```
2023-11-06 11:26:31 Exception in ASGI application
2023-11-06 11:26:31 Traceback (most recent call last):
2023-11-06 11:26:31   File "/usr/local/lib/python3.9/site-packages/uvicorn/protocols/http/httptools_impl.py", line 426, in run_asgi
2023-11-06 11:26:31     result = await app(  # type: ignore[func-returns-value]
2023-11-06 11:26:31   File "/usr/local/lib/python3.9/site-packages/uvicorn/middleware/proxy_headers.py", line 84, in __call__
2023-11-06 11:26:31     return await self.app(scope, receive, send)
2023-11-06 11:26:31   File "/usr/local/lib/python3.9/site-packages/fastapi/applications.py", line 292, in __call__
2023-11-06 11:26:31     await super().__call__(scope, receive, send)
2023-11-06 11:26:31   File "/usr/local/lib/python3.9/site-packages/starlette/applications.py", line 122, in __call__
2023-11-06 11:26:31     await self.middleware_stack(scope, receive, send)
2023-11-06 11:26:31   File "/usr/local/lib/python3.9/site-packages/ddtrace/contrib/asgi/middleware.py", line 275, in __call__
2023-11-06 11:26:31     reraise(exc_type, exc_val, exc_tb)
2023-11-06 11:26:31   File "/usr/local/lib/python3.9/site-packages/six.py", line 719, in reraise
2023-11-06 11:26:31     raise value
2023-11-06 11:26:31   File "/usr/local/lib/python3.9/site-packages/ddtrace/contrib/asgi/middleware.py", line 270, in __call__
2023-11-06 11:26:31     return await self.app(scope, receive, wrapped_send)
2023-11-06 11:26:31   File "/usr/local/lib/python3.9/site-packages/starlette/middleware/errors.py", line 184, in __call__
2023-11-06 11:26:31     raise exc
2023-11-06 11:26:31   File "/usr/local/lib/python3.9/site-packages/starlette/middleware/errors.py", line 162, in __call__
2023-11-06 11:26:31     await self.app(scope, receive, _send)
2023-11-06 11:26:31   File "/usr/local/lib/python3.9/site-packages/starlette/middleware/cors.py", line 83, in __call__
2023-11-06 11:26:31     await self.app(scope, receive, send)
2023-11-06 11:26:31   File "/usr/local/lib/python3.9/site-packages/starlette/middleware/exceptions.py", line 79, in __call__
2023-11-06 11:26:31     raise exc
2023-11-06 11:26:31   File "/usr/local/lib/python3.9/site-packages/starlette/middleware/exceptions.py", line 68, in __call__
2023-11-06 11:26:31     await self.app(scope, receive, sender)
2023-11-06 11:26:31   File "/usr/local/lib/python3.9/site-packages/fastapi/middleware/asyncexitstack.py", line 20, in __call__
2023-11-06 11:26:31     raise e
2023-11-06 11:26:31   File "/usr/local/lib/python3.9/site-packages/fastapi/middleware/asyncexitstack.py", line 17, in __call__
2023-11-06 11:26:31     await self.app(scope, receive, send)
2023-11-06 11:26:31   File "/usr/local/lib/python3.9/site-packages/starlette/routing.py", line 718, in __call__
2023-11-06 11:26:31     await route.handle(scope, receive, send)
2023-11-06 11:26:31   File "/usr/local/lib/python3.9/site-packages/starlette/routing.py", line 276, in handle
2023-11-06 11:26:31     await self.app(scope, receive, send)
2023-11-06 11:26:31   File "/usr/local/lib/python3.9/site-packages/starlette/routing.py", line 66, in app
2023-11-06 11:26:31     response = await func(request)
2023-11-06 11:26:31   File "/usr/local/lib/python3.9/site-packages/fastapi/routing.py", line 273, in app
2023-11-06 11:26:31     raw_response = await run_endpoint_function(
2023-11-06 11:26:31   File "/usr/local/lib/python3.9/site-packages/fastapi/routing.py", line 190, in run_endpoint_function
2023-11-06 11:26:31     return await dependant.call(**values)
2023-11-06 11:26:31   File "/usr/local/lib/python3.9/site-packages/strawberry/fastapi/router.py", line 227, in handle_http_post
2023-11-06 11:26:31     return await self.run(
2023-11-06 11:26:31   File "/usr/local/lib/python3.9/site-packages/strawberry/http/async_base_view.py", line 176, in run
2023-11-06 11:26:31     result = await self.execute_operation(
2023-11-06 11:26:31   File "/usr/local/lib/python3.9/site-packages/strawberry/http/async_base_view.py", line 115, in execute_operation
2023-11-06 11:26:31     return await self.schema.execute(
2023-11-06 11:26:31   File "/usr/local/lib/python3.9/site-packages/strawberry/schema/schema.py", line 256, in execute
2023-11-06 11:26:31     result = await execute(
2023-11-06 11:26:31   File "/usr/local/lib/python3.9/site-packages/strawberry/schema/execute.py", line 86, in execute
2023-11-06 11:26:31     async with extensions_runner.operation():
2023-11-06 11:26:31   File "/usr/local/lib/python3.9/site-packages/strawberry/extensions/context.py", line 191, in __aenter__
2023-11-06 11:26:31     await self.run_hooks_async()
2023-11-06 11:26:31   File "/usr/local/lib/python3.9/site-packages/strawberry/extensions/context.py", line 175, in run_hooks_async
2023-11-06 11:26:31     await hook.initialized_hook.__anext__()  # type: ignore[union-attr]
2023-11-06 11:26:31   File "/src/src/routers/graphql.py", line 78, in on_operation
2023-11-06 11:26:31     raise StrawberryGraphQLError(message="Query too large")
2023-11-06 11:26:31 strawberry.exceptions.StrawberryGraphQLError: Query too large
```

## System Information

 - Operating system: Mac with Docker containers
 - Strawberry version (if applicable): `strawberry-graphql[fastapi] == 0.209.2`

## Upvote & Fund

- We're using [Polar.sh](https://polar.sh/strawberry-graphql) so you can upvote and help fund this issue.
- We receive the funding once the issue is completed & confirmed by you.
- Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/strawberry-graphql/strawberry/issues/3204">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/strawberry-graphql/strawberry/issues/3204/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by DoctorJohn**
This PR brings schema extension error handling in line with field extension and resolver error handling.

## Description

This PR alters schema extension error handling so that any error raised from within schema
extensions will abort the operation and is added to the execution result.

Before, most errors (including Strawberry/graphql-core GraphQLErrors) would cause an internal server error.

Furthermore, these changes enable extensions (e.g. the one described in #3204) to abort operations early in a requests lifecycle which was not easily possible before.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [x] New feature
- [x] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

* Fixes #3204
* Addresses [this comment](https://github.com/strawberry-graphql/strawberry/pull/1324/files#r1389356058).

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by erikwrede** [comment]
I am in favor of addressing #3204 in the same go, as the changing of the current mechanism might create some breakage for existing web views which we could mitigate by integrating the changes required from #3204. Should only be a few lines after all.

**Comment by DoctorJohn** [comment]
> I am in favor of addressing #3204 in the same go, as the changing of the current mechanism might create some breakage for existing web views which we could mitigate by integrating the changes required from #3204. Should only be a few lines after all.

Just to make sure we're on the same page: In its current state this PR does not change the current mechanism, since it's only removing an unnecessary "fallback" `except` block wrapping a single call to the graphql-core `parse` function.

**Comment by botberry** [comment]
Thanks for adding the `RELEASE.md` file!



Here's a preview of the changelog:

---

Starting with this release, any error raised from within schema
extensions will abort the operation and is returned to the client.

This corresponds to the way we already handle field extension errors
and resolver errors.

This is particular useful for schema extensions performing checks early
in the request lifecycle, for example:

```python
class MaxQueryLengthExtension(SchemaExtension):
    MAX_QUERY_LENGTH = 8192

    async def on_operation(self):
        if len(self.execution_context.query) > self.MAX_QUERY_LENGTH:
            raise StrawberryGraphQLError(message="Query too large")
        yield
```



Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to @NucleonJohn for the PR 👏

Get it here 👉 https://beta.strawberry.rocks/release/(next)
```

File: strawberry/schema/execute.py [hunk 1]
```diff
@@ -88,77 +88,78 @@ async def execute(
...
-        # Strawberry we are validating it at initialisation time instead
-        if not execution_context.query:
-            raise MissingQueryError()
-
-        async with extensions_runner.parsing():
-            try:
-                if not execution_context.graphql_document:
-                    execution_context.graphql_document = parse_document(
-                        execution_context.query, **execution_context.parse_options
+    try:
```
Comment (DoctorJohn) [inline strawberry/schema/execute.py:91]: This whole diff is essentialy just adding the try/except statement.

File: strawberry/schema/execute.py [hunk 2]
```diff
@@ -88,77 +88,78 @@ async def execute(
...
+                    if result.errors:
+                        execution_context.errors = result.errors
+
+                        # Run the `Schema.process_errors` function here before
+                        # extensions have a chance to modify them (see the MaskErrors
+                        # extension). That way we can log the original errors but
+                        # only return a sanitised version to the client.
+                        process_errors(result.errors, execution_context)
+
+    except (MissingQueryError, InvalidOperationTypeError) as e:
```
Comment (DoctorJohn) [inline strawberry/schema/execute.py:152]: These two exceptions get special handling by our view integrations and enables them to return appropriate status codes. For context:

https://github.com/strawberry-graphql/strawberry/blob/d6bd38daf7c381169d65e6be92666ff1109ec64e/strawberry/http/sync_base_view.py#L198-L209

Comment (erikwrede) [reply strawberry/schema/execute.py:152]: I understand that this solves the problem for now. However, I doubt that this is a long term solution due to the specificity of re-raising only specific exceptions. We should try to abstract that into an exception -> status code or modified response or graphQL error "middleware/exception filter" as a follow up to this PR.

**Comment by botberry** [comment]
# Pre-release
:wave:

Releasing commit [4efd89300e98ac163fdc3e074a30014fce69093e] to PyPi as pre-release! :package:

**Comment by botberry** [comment]
# Pre-release
:wave:

Pre-release **0.214.0.dev.1701082152** [4efd89300e98ac163fdc3e074a30014fce69093e] has been released on PyPi! :rocket:
You can try it by doing:

```shell
poetry add strawberry-graphql==0.214.0.dev.1701082152
```

**Comment by imconnorngl** [comment]
After testing this compared to the functionality of https://github.com/strawberry-graphql/strawberry/pull/3256. It seems like this PR is more verbose than I'd like in best case, as seen below

Before:
image
After:
image

File: strawberry/schema/execute.py (hunk 2 above)
Comment (patrick91) [reply strawberry/schema/execute.py:152]: something like an `OperationError`? 😊

**Comment by patrick91** [comment]
I've merged main into this, but something has changed 😊 I'll take a proper look at this in the following days

File: strawberry/extensions/context.py [hunk 3]
```diff
@@ -166,14 +168,17 @@ def __enter__(self) -> None:
...
+            for hook in self.hooks:
+                if hook.is_async:
+                    raise RuntimeError(
+                        f"SchemaExtension hook {hook.extension}.{self.HOOK_NAME} "
+                        "failed to complete synchronously."
+                    )
+                else:
+                    self.exit_stack.enter_context(hook.hook())  # type: ignore
+        except Exception as e:
+            self.pending_exceptions.append(e)
```
Comment (patrick91) [inline strawberry/extensions/context.py:181]: we could potentially raise an exception group, if there's a backport for it 😊

File: tests/schema/extensions/test_extensions.py [hunk 4]
```diff
@@ -721,7 +725,7 @@ def ping(self) -> str:
 @pytest.mark.parametrize(
     ("failing_hook", "expected_hooks"),
     (
-        ("on_operation_start", set()),
+        ("on_operation_start", {2, 3, 4, 5, 6, 7}),
```
[the remark this answers is not in the document]
Comment (patrick91) [reply tests/schema/extensions/test_extensions.py:728]: mmh, looking at the current failure it looks like what we were trying to do is not working as I though

we can't stop an operation from an extension if we catch all exceptions, unless I'm missing something

File: strawberry/schema/execute.py (hunk 2 above)
Comment (patrick91) [reply strawberry/schema/execute.py:152]: let's do in another PR/refactoring 😊

**Pull request merged**